### PR TITLE
Shuffle toolbar button

### PR DESCRIPTION
**Pull request by Ben2356**
Added a shuffle toolbar button which will provide shuffle functionality if it is supported for your current playlist/queue and by the music application. The button and tooltip text will change depending on if you are shuffling or not. If the shuffle functionality is not supported by the app or for the playlist/queue it will be greyed out.

Shuffle button in action:

### CODE REVIEW

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -118,6 +118,8 @@
     <string name="MUSIC_QUEUE_EMPTY">&lt;Empty Queue&gt;</string>
     <string name="MUSIC_SKIP_PREVIOUS">Back</string>
     <string name="MUSIC_SKIP_NEXT">Next</string>
+    <string name="MUSIC_TURN_SHUFFLE_ON">Turn Shuffle On</string>
```
[the remark this answers is not in the document]
Comment (Ben2356) [reply app/src/main/res/values/strings.xml:121]: Renamed and moved the keys (spatial location in file) for all languages

**Comment by hufman** [comment]
It looks really good! It works great for Spotify, delightfully responsive and updates instantly no matter which screen toggles the setting. However, I couldn't get any GenericMusicAppController apps (I tried my favorites: DSub, Plex, Rocket Player, Soundcloud, VLC) to show the shuffle button as enabled. I tried digging in to YouTube Music like you mentioned, and it also didn't claim to support SET_SHUFFLE_MODE. What other apps did you find that supported shuffle mode?
I feel like it's not a commonly-supported extension, because Android Auto more prominently displays the app's custom actions on its main screen. I think the Shuffle button should be hidden when Shuffle is not supported (with the same imageId=0/setVisible trick as before), because IDrive5+ only shows 6 toolbar buttons and so the Skip Next button is scrolled off the bottom.

One more tiny tweak: MusicAction.SET_SHUFFLE_MODE should be changed to `ACTION_SET_SHUFFLE_MODE` because `ACTION_SET_SHUFFLE_MODE_ENABLED` is apparently deprecated, and it would line up more exactly to the setShuffleMode function call that would be enabled if supported.

Thank you so much for working on this!

**Comment by Ben2356** [comment]
> 
> 
> It looks really good! It works great for Spotify, delightfully responsive and updates instantly no matter which screen toggles the setting. However, I couldn't get any GenericMusicAppController apps (I tried my favorites: DSub, Plex, Rocket Player, Soundcloud, VLC) to show the shuffle button as enabled. I tried digging in to YouTube Music like you mentioned, and it also didn't claim to support SET_SHUFFLE_MODE. What other apps did you find that supported shuffle mode?
> I feel like it's not a commonly-supported extension, because Android Auto more prominently displays the app's custom actions on its main screen. I think the Shuffle button should be hidden when Shuffle is not supported (with the same imageId=0/setVisible trick as before), because IDrive5+ only shows 6 toolbar buttons and so the Skip Next button is scrolled off the bottom.
> 
> One more tiny tweak: MusicAction.SET_SHUFFLE_MODE should be changed to `ACTION_SET_SHUFFLE_MODE` because `ACTION_SET_SHUFFLE_MODE_ENABLED` is apparently deprecated, and it would line up more exactly to the setShuffleMode function call that would be enabled if supported.
> 
> Thank you so much for working on this!

Thank you! I thought it worked with Youtube Music but apparently I'm wrong and must have had a bug that was displaying the button as enabled when I was testing it. I'm actually not sure of any apps that support SET_SHUFFLE_MODE, I haven't used any other music apps other than Spotify and Youtube Music. Yeah it must not be that common then, I made some changes so now if the app does not support shuffling the button will be hidden. I also tested going between Spotify and Youtube Music and it behaves correctly when the app is switched. I did noticed that when I tried to do the imageId=0  trick that once the imageId is set to 0 I couldn't get it to be assigned to another icon so it would end up blank when it was supported again but the setVisible method worked perfectly so just using that instead. Also updated the SET_SHUFFLE_MODE to the ACTION_SHUFFLE_MODE enum.

**Comment by hufman** [comment]
Ah ha, you discovered a strange quirk of IDrive5+: It actually validates the imageId against the contents of the icon pack, and will raise an exception if it isn't valid. However, for IDrive4, it seems to be the best way to make a toolbar icon become empty while still taking up space, because the toolbar looks incorrect when it is missing buttons.

Does the `redrawShuffleButton` behave correctly if you wrap the hiding logic in a try/catch block like the original hiding code?

**Comment by Ben2356** [comment]
> 
> 
> Ah ha, you discovered a strange quirk of IDrive5+: It actually validates the imageId against the contents of the icon pack, and will raise an exception if it isn't valid. However, for IDrive4, it seems to be the best way to make a toolbar icon become empty while still taking up space, because the toolbar looks incorrect when it is missing buttons.
> 
> Does the `redrawShuffleButton` behave correctly if you wrap the hiding logic in a try/catch block like the original hiding code?

Ah that's interesting, didn't know that! Yeah I did add the try catch inside of the redrawShuffleButton method using the original code and that's when I noticed the icon not being set correctly after it was set to 0, the button spot where it would be would be a blank spot and wouldn't load any sort of icon. I'm running on iDrive 6 so not sure if that was something that was changed maybe?

**Comment by hufman** [comment]
Oh very very interesting!! You have discovered a new buggy behavior of the car itself haha. It seems that, even though the car raised an exception (and thus aborted updating the value for the idempotent layer), it still clears the icon, and so the idempotent layer has the chance of not setting the icon when it becomes supported again, leaving the icon clear.
The following snippet worked for me:
```
	private fun redrawShuffleButton() {
		if (controller.isSupportedAction(MusicAction.SET_SHUFFLE_MODE)) {
			if(controller.isShuffling()) {
				shuffleButton.getTooltipModel()?.asRaDataModel()?.value = L.MUSIC_TURN_SHUFFLE_OFF
				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = IMAGEID_SHUFFLE_ON
			}
			else {
				shuffleButton.getTooltipModel()?.asRaDataModel()?.value = L.MUSIC_TURN_SHUFFLE_ON
				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = IMAGEID_SHUFFLE_OFF
			}
			shuffleButton.setVisible(true)
		} else {
			try {
				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = 0
			} catch (e: BMWRemoting.ServiceException) {
				shuffleButton.setVisible(false)
				// the car has cleared the icon even though it threw an exception
				// so set the icon to a valid imageId again
				// to make sure the idempotent layer properly sets the icon in the future
				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = IMAGEID_SONG
			}
		}
	}
```

I'll merge it as soon as this change is in (which might require updating tests). Again, thank you so much!

**Comment by Ben2356** [comment]
> 
> 
> Oh very very interesting!! You have discovered a new buggy behavior of the car itself haha. It seems that, even though the car raised an exception (and thus aborted updating the value for the idempotent layer), it still clears the icon, and so the idempotent layer has the chance of not setting the icon when it becomes supported again, leaving the icon clear.
> The following snippet worked for me:
> 
> ```
> 	private fun redrawShuffleButton() {
> 		if (controller.isSupportedAction(MusicAction.SET_SHUFFLE_MODE)) {
> 			if(controller.isShuffling()) {
> 				shuffleButton.getTooltipModel()?.asRaDataModel()?.value = L.MUSIC_TURN_SHUFFLE_OFF
> 				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = IMAGEID_SHUFFLE_ON
> 			}
> 			else {
> 				shuffleButton.getTooltipModel()?.asRaDataModel()?.value = L.MUSIC_TURN_SHUFFLE_ON
> 				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = IMAGEID_SHUFFLE_OFF
> 			}
> 			shuffleButton.setVisible(true)
> 		} else {
> 			try {
> 				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = 0
> 			} catch (e: BMWRemoting.ServiceException) {
> 				shuffleButton.setVisible(false)
> 				// the car has cleared the icon even though it threw an exception
> 				// so set the icon to a valid imageId again
> 				// to make sure the idempotent layer properly sets the icon in the future
> 				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = IMAGEID_SONG
> 			}
> 		}
> 	}
> ```
> 
> I'll merge it as soon as this change is in (which might require updating tests). Again, thank you so much!
> 
> Edit: Actually, if shuffle isn't supported and on IDrive5+, this will keep toggling the imageId between 0 and the placeholder... Some conditional will need to be added against this.

So I came up with this solution here, basically just uses a class member var to determine if we already toggled the imageId so we know not to do that workflow again for iDrive5+ but for iDrive4 it will still set the imageId to be 0. I don't have any iDrive4 systems to test on but does that throw an exception too for when the imageId is set to 0? If it doesn't then this should work out I believe otherwise I don't think there is a way around the constant toggling of the imageId that I'm able to think of at least. 

`private fun redrawShuffleButton() {
		if (controller.isSupportedAction(MusicAction.SET_SHUFFLE_MODE)) {
			if(controller.isShuffling()) {
				shuffleButton.getTooltipModel()?.asRaDataModel()?.value = L.MUSIC_TURN_SHUFFLE_OFF
				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = IMAGEID_SHUFFLE_ON
			} else {
				shuffleButton.getTooltipModel()?.asRaDataModel()?.value = L.MUSIC_TURN_SHUFFLE_ON
				shuffleButton.getImageModel()?.asImageIdModel()?.imageId = IMAGEID_SHUFFLE_OFF
			}
			shuffleButton.setVisible(true)
		} else {
			if(isNewerIDrive) {
				shuffleButton.setVisible(false)
			} else {
				try {
					shuffleButton.getImageModel()?.asImageIdModel()?.imageId = 0
				} catch (e: BMWRemoting.ServiceException) {
					isNewerIDrive = true
					shuffleButton.setVisible(false)
					
					// the car has cleared the icon even though it threw an exception
					// so set the icon to a valid imageId again
					// to make sure the idempotent layer properly sets the icon in the future
					shuffleButton.getImageModel()?.asImageIdModel()?.imageId = IMAGEID_SONG
				}
			}
		}`

**Comment by hufman** [comment]
That looks perfect!

**Comment by Ben2356** [comment]
> 
> 
> That looks perfect!

Awesome, just added in those changes and also added some more unit tests I realized I was missing!

**Comment by hufman** [comment]
Thank you so much!!

**Pull request merged**